# Used Cars in eBay - Germany Classifieds 

This project will analyze ads for used cars on eBay Kleinanzeigen, a classifieds section of the German eBay website. The original dataset can be found on Data.World [here](https://data.world/data-society/used-cars-data).

The dataset used is a sampled 50,000 data points for a smoother compilation in the hosts server. The table below lists the data dictionary. 

| Data                | Dictionary                                                                  |
| ------------------- | --------------------------------------------------------------------------- |
| dateCrawled         |  When this ad was first crawled. All field values are taken from this date. |
| name                |  Name of the car.                                                           |
| seller              |  Whether the seller is private or a dealer.                                 |
| offerType           |  The type of listing                                                        |
| price               |  The price on the ad to sell the car.                                       |
| abtest              |  Whether the listing is included in an A/B test.                            |
| vehicleType         |  The vehicle Type.                                                          |
| yearOfRegistration  |  The year in which the car was first registered.                            |
| gearbox             |  The transmission type.                                                     |
| powerPS             |  The power of the car in PS.                                                |
| model               |  The car model name.                                                        |
| kilometer           |  How many kilometers the car has driven.                                    |
| monthOfRegistration |  The month in which the car was first registered.                           |
| fuelType            |  What type of fuel the car uses.                                            |
| brand               |  The brand of the car.                                                      |
| notRepairedDamage   |  If the car has a damage which is not yet repaired.                         |
| dateCreated         |  The date on which the eBay listing was created.                            |
| nrOfPictures        |  The number of pictures in the ad.                                          |
| postalCode          |  The postal code for the location of the vehicle.                           |
| lastSeenOnline      |  When the crawler saw this ad last online.                                  |

The goal of the project is to practice data cleaning primarily using pandas.  

In [303]:
import pandas as pd
import numpy as np 

Reading the autos.csv using pandas

In [304]:
autos = pd.read_csv("autos.csv", encoding = "Latin-1")
pd.set_option('display.float_format', lambda x: '%.2f' % x)

In [305]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [306]:
autos.describe(include=['O'])

dateCrawled         name  seller offerType  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-04-02 15:49:30  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   

       vehicleType  gearbox  model   odometer fuelType       brand  \
count        44905    47320  47242      50000    45518       50000   
unique           8        2    245         13        7          40   
top      limousine  manuell   golf  150,000km   benzin  volkswagen   
freq         12859    36993   4024      32424    30107       10687   

       notRepairedDamage          dateCreated             lastSeen  
count              40171                50000                50000  
unique                 2                   76                39481  
top                 nein  2016-04-03 00:00:00  2016-04-07 06:17:27  
freq               35232                 1946                    8

## Initial comments on string columns:

- Some data values are in german, as expected but a little harder to deal with.
- `dateCrawled`, `dateCreated`, `lastSeen` should be converted to date dtype. 
- `price` should be converted to numeric dtype. 
- Unsure about the `abtest` columns usefulness. 
- `odometer` should be converted to numeric dtype. 

In [307]:
autos.describe()

yearOfRegistration  powerPS  monthOfRegistration  nrOfPictures  \
count            50000.00 50000.00             50000.00      50000.00   
mean              2005.07   116.36                 5.72          0.00   
std                105.71   209.22                 3.71          0.00   
min               1000.00     0.00                 0.00          0.00   
25%               1999.00    70.00                 3.00          0.00   
50%               2003.00   105.00                 6.00          0.00   
75%               2008.00   150.00                 9.00          0.00   
max               9999.00 17700.00                12.00          0.00   

       postalCode  
count    50000.00  
mean     50813.63  
std      25779.75  
min       1067.00  
25%      30451.00  
50%      49577.00  
75%      71540.00  
max      99998.00

## Initial comments on numeric columns:

- `yearOfRegistration` has outliers, minimum year of 1000 and max of 9999. 
- `power` cannot possibly be 17700 PS.
- `postalCode` can be verified if need be. 

Changing the column names to snake_case and changing some names for better comprehension. 

In [308]:
column_names = autos.columns.tolist()
print(column_names)

['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest', 'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model', 'odometer', 'monthOfRegistration', 'fuelType', 'brand', 'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode', 'lastSeen']


In [309]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest', 'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model', 'odometer', 'registration_month', 'fuel_type', 'brand', 'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code', 'last_seen']

In [310]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [311]:
autos.head(5)

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Dropping columns with values of little/no interest given that they have no descriptive statistics. Removing the following [`seller`, `offerType`] Also changing the dtypes for odometer reading and price. 

In [312]:
autos.loc[:,"price"]=autos.loc[:,"price"].map(lambda x: x.strip("$").replace(",","")).astype("int64")
autos.loc[:,"odometer"]=autos.loc[:,"odometer"].map(lambda x: x.strip("km,").replace(",","")).astype("int64")
autos.rename(columns={"price": "price_usd", "odometer":"odometer_km"}, inplace=True)

In [313]:
autos.dtypes


date_crawled          object
name                  object
seller                object
offer_type            object
price_usd              int64
abtest                object
vehicle_type          object
registration_year      int64
gearbox               object
power_ps               int64
model                 object
odometer_km            int64
registration_month     int64
fuel_type             object
brand                 object
unrepaired_damage     object
ad_created            object
nr_of_pictures         int64
postal_code            int64
last_seen             object
dtype: object

In [314]:
autos.describe()

price_usd  registration_year  power_ps  odometer_km  \
count    50000.00           50000.00  50000.00     50000.00   
mean      9840.04            2005.07    116.36    125732.70   
std     481104.38             105.71    209.22     40042.21   
min          0.00            1000.00      0.00      5000.00   
25%       1100.00            1999.00     70.00    125000.00   
50%       2950.00            2003.00    105.00    150000.00   
75%       7200.00            2008.00    150.00    150000.00   
max   99999999.00            9999.00  17700.00    150000.00   

       registration_month  nr_of_pictures  postal_code  
count            50000.00        50000.00     50000.00  
mean                 5.72            0.00     50813.63  
std                  3.71            0.00     25779.75  
min                  0.00            0.00      1067.00  
25%                  3.00            0.00     30451.00  
50%                  6.00            0.00     49577.00  
75%                  9.00            0.00     71540.00  
max                 12.00            0.00     99998.00

In [315]:
autos["price_usd"].unique().shape
autos["price_usd"].describe()

count      50000.00
mean        9840.04
std       481104.38
min            0.00
25%         1100.00
50%         2950.00
75%         7200.00
max     99999999.00
Name: price_usd, dtype: float64

In [316]:
autos["price_usd"].value_counts(sort=False).sort_index().tail(20)

197000      1
198000      1
220000      1
250000      1
259000      1
265000      1
295000      1
299000      1
345000      1
350000      1
999990      1
999999      2
1234566     1
1300000     1
3890000     1
10000000    1
11111111    2
12345678    3
27322222    1
99999999    1
Name: price_usd, dtype: int64

In [317]:
autos["registration_year"][autos["registration_year"].between(1900,2020)]

0        2004
1        1997
2        2009
3        2007
4        2003
         ... 
49995    2011
49996    1996
49997    2014
49998    2013
49999    1996
Name: registration_year, Length: 49976, dtype: int64

## Added comments/solutions on numeric columns:

- `price_usd` has outliers for a free car and one worth over 99 million. Post some analysis, any outliers past $(100,350000) will be removed.
- `registration_year` will also be truncated to be (1900,2020) leaving us with 49976 datapoints. 

In [318]:
autos_trimmed = autos[(autos['price_usd'].between(100,350000)) & (autos['registration_year'].between(1900,2020))]

In [319]:
autos_trimmed['price_usd'].value_counts(sort=True, bins=10).sort_index()

(-249.901, 35090.0]     47697
(35090.0, 70080.0]        426
(70080.0, 105070.0]        46
(105070.0, 140060.0]       15
(140060.0, 175050.0]        8
(175050.0, 210040.0]        5
(210040.0, 245030.0]        1
(245030.0, 280020.0]        3
(280020.0, 315010.0]        2
(315010.0, 350000.0]        2
Name: price_usd, dtype: int64

In [320]:
autos_trimmed["registration_year"].value_counts(sort=True, bins=10).sort_index()

(1909.89, 1920.9]        2
(1920.9, 1931.8]         3
(1931.8, 1942.7]        10
(1942.7, 1953.6]         7
(1953.6, 1964.5]        72
(1964.5, 1975.4]       271
(1975.4, 1986.3]       534
(1986.3, 1997.2]      6932
(1997.2, 2008.1]     28883
(2008.1, 2019.0]     11491
Name: registration_year, dtype: int64

In [321]:
autos_trimmed

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offer_type  price_usd   abtest vehicle_type  registration_year  \
0      privat    Angebot       5000  control          bus               2004   
1      privat    Angebot       8500  control    limousine               1997   
2      privat    Angebot       8990     test    limousine               2009   
3      privat    Angebot       4350  control   kleinwagen               2007   
4      privat    Angebot       1350     test        kombi               2003   
...       ...        ...        ...      ...          ...                ...   
49995  privat    Angebot      24900  control    limousine               2011   
49996  privat    Angebot       1980  control       cabrio               1996   
49997  privat    Angebot      13200     test       cabrio               2014   
49998  privat    Angebot      22900  control        kombi               2013   
49999  privat    Angebot       1250  control    limousine               1996   

         gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0        manuell       158  andere       150000                   3       lpg   
1      automatik       286     7er       150000                   6    benzin   
2        manuell       102    golf        70000                   7    benzin   
3      automatik        71  fortwo        70000                   6    benzin   
4        manuell         0   focus       150000                   7    benzin   
...          ...       ...     ...          ...                 ...       ...   
49995  automatik       239      q5       100000                   1    diesel   
49996    manuell        75   astra       150000                   5    benzin   
49997  automatik        69     500         5000                  11    benzin   
49998    manuell       150      a3        40000                  11    diesel   
49999    manuell       101  vectra       150000                   1    benzin   

            brand unrepaired_damage           ad_created  nr_of_pictures  \
0         peugeot              nein  2016-03-26 00:00:00               0   
1             bmw              nein  2016-04-04 00:00:00               0   
2      volkswagen              nein  2016-03-26 00:00:00               0   
3           smart              nein  2016-03-12 00:00:00               0   
4            ford              nein  2016-04-01 00:00:00               0   
...           ...               ...                  ...             ...   
49995        audi              nein  2016-03-27 00:00:00               0   
49996        opel              nein  2016-03-28 00:00:00               0   
49997        fiat              nein  2016-04-02 00:00:00               0   
49998        audi              nein  2016-03-08 00:00:00               0   
49999        opel              nein  2016-03-13 00:00:00               0   

       postal_code            last_seen  
0            79588  2016-04-06 06:45:54  
1            71034  2016-04-06 14:45:08  
2            35394  2016-04-06 20:15:37  
3            33729  2016-0

## Aggregation by mean price 

Now that the data has been cleaned, some aggregation will be performed for the `brand` column to identify the brands with the highest mean values. 

In [322]:
brand_dict = autos_trimmed["brand"].value_counts().to_dict()

In [323]:
autos_trimmed.head(5)

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price_usd   abtest vehicle_type  registration_year  \
0  privat    Angebot       5000  control          bus               2004   
1  privat    Angebot       8500  control    limousine               1997   
2  privat    Angebot       8990     test    limousine               2009   
3  privat    Angebot       4350  control   kleinwagen               2007   
4  privat    Angebot       1350     test        kombi               2003   

     gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere       150000                   3       lpg   
1  automatik       286     7er       150000                   6    benzin   
2    manuell       102    golf        70000                   7    benzin   
3  automatik        71  fortwo        70000                   6    benzin   
4    manuell         0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [327]:
brand_mean_price_usd = {}

for brand in brand_dict.keys():
    brand_df = autos_trimmed[autos_trimmed.loc[:,"brand"] == brand][["brand","price_usd"]]
    brand_mean_price_usd[brand] = round(brand_df.loc[:,"price_usd"].mean(), 2)
    
print(brand_mean_price_usd)

{'volkswagen': 5366.25, 'bmw': 8310.14, 'opel': 2971.61, 'mercedes_benz': 8570.77, 'audi': 9259.51, 'ford': 3756.99, 'renault': 2450.9, 'peugeot': 3086.93, 'fiat': 2815.64, 'seat': 4353.15, 'skoda': 6394.31, 'nissan': 4681.94, 'mazda': 4075.32, 'smart': 3538.34, 'citroen': 3777.85, 'toyota': 5148.0, 'hyundai': 5416.23, 'sonstige_autos': 12640.21, 'volvo': 4911.68, 'mini': 10566.82, 'mitsubishi': 3429.87, 'honda': 4010.47, 'kia': 5923.29, 'alfa_romeo': 4054.47, 'suzuki': 4166.77, 'porsche': 46764.2, 'chevrolet': 6692.6, 'chrysler': 3539.92, 'dacia': 5897.74, 'daihatsu': 1641.26, 'jeep': 11590.21, 'subaru': 4019.07, 'land_rover': 18934.27, 'saab': 3183.49, 'daewoo': 1093.6, 'jaguar': 11844.04, 'trabant': 1843.54, 'rover': 1586.49, 'lancia': 3240.7, 'lada': 2647.72}


## Cars with highest mean-price

In [342]:
brand_price = pd.DataFrame.from_dict(brand_mean_price_usd, orient='index', columns=["price"])
brand_price["price"].sort_values(ascending=False).head(5)

porsche          46764.20
land_rover       18934.27
sonstige_autos   12640.21
jaguar           11844.04
jeep             11590.21
Name: price, dtype: float64